Mathlib/RingTheory/AlgebraicIndependent/TranscendenceBasis.lean

```diff
@@ -125,3 +233,204 @@ theorem IsTranscendenceBasis.isAlgebraic_field {F E : Type*} {x : ι → E}
+  refine ⟨fun le ↦ ?_, (·.lift_cardinalMk_le_trdeg)⟩
+  have ⟨s, hsx, hs⟩ := exists_isTranscendenceBasis_subset inj (range x)
+  rw [← hs.cardinalMk_eq_trdeg] at le
+  obtain rfl := (finite_range x).eq_of_subset_of_encard_le hsx
+    (toENat.monotone' <| lift_le.{u}.mp <| mk_range_le_lift.trans le)
+  have ⟨_, inj⟩ := lift_mk_le'.mp le
+  have := surjective_onto_range.bijective_of_nat_card_le (Nat.card_le_card_of_injective _ inj)
+  exact .of_subtype_range (fun _ _ ↦ (this.1 <| Subtype.ext ·)) hs.1
+
+theorem cardinalMk_eq_trdeg_iff_of_finite {ι : Type w} [Finite ι] (x : ι → A)
+    [Algebra.IsAlgebraic (adjoin R (range x)) A] :
+    #ι = trdeg R A ↔ IsTranscendenceBasis R x := by
+  rw [← lift_cardinalMk_eq_trdeg_iff_of_finite R x inj, lift_id, lift_id]
+
+theorem lift_cardinalMk_eq_trdeg_iff [Algebra.IsAlgebraic (adjoin R (range x)) A]
+    (fin : trdeg R A < ℵ₀) : lift.{w} #ι = lift.{u} (trdeg R A) ↔ IsTranscendenceBasis R x := by
+  have := notrivial_of_adjoin_range R x
+  refine ⟨fun h ↦ ?_, (·.lift_cardinalMk_eq_trdeg)⟩
+  have := mk_lt_aleph0_iff.mp (lift_lt_aleph0.mp (lift_aleph0 ▸ h ▸ lift_lt.{_, u}.mpr fin))
+  exact (lift_cardinalMk_eq_trdeg_iff_of_finite R x inj).mp h
+
+theorem cardinalMk_eq_trdeg_iff {ι : Type w} (x : ι → A)
+    [Algebra.IsAlgebraic (adjoin R (range x)) A] (fin : trdeg R A < ℵ₀) :
+    #ι = trdeg R A ↔ IsTranscendenceBasis R x := by
+  rw [← lift_cardinalMk_eq_trdeg_iff R x inj fin, lift_id, lift_id]
+
+end Algebra.IsAlgebraic
+
+namespace AlgebraicIndependent
+
+variable [Nontrivial R] [NoZeroDivisors A]
+
+theorem lift_cardinalMk_eq_trdeg_iff_of_finite [Finite ι] (hx : AlgebraicIndependent R x) :
+    lift.{w} #ι = lift.{u} (trdeg R A) ↔ IsTranscendenceBasis R x := by
+  have ⟨s, hxs, hs⟩ := exists_isTranscendenceBasis_superset hx.to_subtype_range
+  refine ⟨fun h ↦ ?_, (·.lift_cardinalMk_eq_trdeg)⟩
+  rw [← hs.cardinalMk_eq_trdeg, ← mk_range_eq_of_injective hx.injective, lift_inj] at h
+  obtain rfl := (finite_range x).eq_of_subset_of_encard_le' hxs (toENat.monotone' h.ge)
+  exact .of_subtype_range hx.injective hs
+
+theorem cardinalMk_eq_trdeg_iff_of_finite {ι : Type w} [Finite ι] {x : ι → A}
+    (hx : AlgebraicIndependent R x) : #ι = trdeg R A ↔ IsTranscendenceBasis R x := by
+  rw [← lift_cardinalMk_eq_trdeg_iff_of_finite hx, lift_id, lift_id]
+
+theorem lift_cardinalMk_eq_trdeg_iff (hx : AlgebraicIndependent R x) (fin : trdeg R A < ℵ₀) :
+    lift.{w} #ι = lift.{u} (trdeg R A) ↔ IsTranscendenceBasis R x := by
+  refine ⟨fun h ↦ ?_, (·.lift_cardinalMk_eq_trdeg)⟩
+  have := mk_lt_aleph0_iff.mp (lift_lt_aleph0.mp (lift_aleph0 ▸ h ▸ lift_lt.{_, u}.mpr fin))
+  exact (lift_cardinalMk_eq_trdeg_iff_of_finite hx).mp h
+
+theorem cardinalMk_eq_trdeg_iff {ι : Type w} {x : ι → A} (hx : AlgebraicIndependent R x)
+    (fin : trdeg R A < ℵ₀) : #ι = trdeg R A ↔ IsTranscendenceBasis R x := by
+  rw [← lift_cardinalMk_eq_trdeg_iff hx fin, lift_id, lift_id]
```

Comment:
`trdeg` is a kind of `rank`, right? I think all these probably have generalizations to matroids.

Reply:
Yes, if it's 'maximum cardinality of an independent subset', then it's matroidal rank. A lot of API for this should belong in `Matroid.Rank.Cardinality` once it's up and running.

Reply:
Is Cardinality a different file than Cardinal.lean I added in #20888?

Reply:
I meant `Cardinal`, sorry.

Reply:
See #20921, which defines the rank function you want. (Note that you're delegated, so you can merge if it looks good). This should provide a way to define `trdeg` in terms of the matroid.

Reply:
Thanks! A bit busy right now but will review/merge later today. The current definition of trdeg doesn't require NoZeroDivisors, but it's required to get a matroid. So I think what I'll do is to show trdeg is equal to matroid rank in a domain.

Reply:
@apnelson1 The following should be true for an arbitrary matroid M; do you already have them somewhere? If not I'll include them in this PR.
```
theorem finiteRk_iff_cRank_lt_aleph0 : M.FiniteRk ↔ M.cRank < ℵ₀ := by
  rw [finiteRk_iff]
  refine ⟨fun ⟨B, base, fin⟩ ↦ ?_, ?_⟩ <;> sorry

theorem Indep.base_iff_eq_cRank [M.FiniteRk] (ind : M.Indep I) : M.Base I ↔ #I = M.cRank := by
  sorry

theorem Spanning.base_iff_eq_cRank [M.FiniteRk] (h : M.Spanning X) : M.Base X ↔ #X = M.cRank := by
  sorry

theorem Indep.base_iff_eq_cRank_of_lt_aleph0 (lt : #I < ℵ₀) (ind : M.Indep I) :
    M.Base I ↔ #I = M.cRank := by
  sorry

theorem Spanning.base_iff_eq_cRank_of_lt_aleph0 (lt : #X < ℵ₀) (h : M.Spanning X) :
    M.Base X ↔ #X = M.cRank := by
  sorry
```

Reply:
I don't have these, no. They make sense and I'd be happy to have them, although might this be scope creep for this PR?

Reply:
Maybe in another PR and this PR also depends on it.

Reply:
Perhaps a comment is that [this file](https://github.com/apnelson1/Matroid/blob/main/Matroid/Rank/Finite.lean) contains a `FinRk` definition that is a more general version of the `Finite` hypotheses in your lemmas. I can PR that soon if it makes sense to do things that way.

Reply:
(It is basically a leaf file that is ready to go in one PR)